sdk/storage/test/shares/share_file_client_test.cpp

```diff
@@ -0,0 +1,473 @@
+    m_fileClient->BreakLease();
+  }
+
+  TEST_F(FileShareFileClientTest, ConcurrentUpload)
+  {
+    std::vector<uint8_t> fileContent = RandomBuffer(static_cast<std::size_t>(8_MB));
+
+    auto testUploadFromBuffer = [&](int concurrency, int64_t fileSize) {
+      auto fileClient = m_fileShareDirectoryClient->GetFileClient(RandomString());
+
+      Files::Shares::UploadFileFromOptions options;
+      options.ChunkSize = 512_KB;
+      options.Concurrency = concurrency;
+      options.HttpHeaders = GetInterestingHttpHeaders();
+      options.Metadata = RandomMetadata();
+
+      auto res
+          = fileClient.UploadFrom(fileContent.data(), static_cast<std::size_t>(fileSize), options);
+
+      auto properties = *fileClient.GetProperties();
+      EXPECT_EQ(properties.ContentLength, fileSize);
+      EXPECT_EQ(properties.Metadata, options.Metadata);
+      std::vector<uint8_t> downloadContent(static_cast<std::size_t>(fileSize), '\x00');
+      fileClient.DownloadTo(downloadContent.data(), static_cast<std::size_t>(fileSize));
+      EXPECT_EQ(
+          downloadContent,
+          std::vector<uint8_t>(
+              fileContent.begin(), fileContent.begin() + static_cast<std::size_t>(fileSize)));
+    };
+
+    auto testUploadFromFile = [&](int concurrency, int64_t fileSize) {
+      auto fileClient = m_fileShareDirectoryClient->GetFileClient(RandomString());
+
+      Files::Shares::UploadFileFromOptions options;
+      options.ChunkSize = 512_KB;
+      options.Concurrency = concurrency;
+      options.HttpHeaders = GetInterestingHttpHeaders();
+      options.Metadata = RandomMetadata();
+
+      std::string tempFilename = RandomString();
+      {
+        Azure::Storage::Details::FileWriter fileWriter(tempFilename);
+        fileWriter.Write(fileContent.data(), fileSize, 0);
+      }
+
+      auto res = fileClient.UploadFrom(tempFilename, options);
+
+      auto properties = *fileClient.GetProperties();
+      EXPECT_EQ(properties.ContentLength, fileSize);
+      EXPECT_EQ(properties.Metadata, options.Metadata);
+      std::vector<uint8_t> downloadContent(static_cast<std::size_t>(fileSize), '\x00');
+      fileClient.DownloadTo(downloadContent.data(), static_cast<std::size_t>(fileSize));
+      EXPECT_EQ(
+          downloadContent,
+          std::vector<uint8_t>(
+              fileContent.begin(), fileContent.begin() + static_cast<std::size_t>(fileSize)));
+
+      DeleteFile(tempFilename);
+    };
+
+    std::vector<std::future<void>> futures;
+    for (int c : {1, 2, 5})
+    {
+      for (int64_t l : {0ULL, 512ULL, 1_KB, 4_KB, 1_MB, 4_MB + 512})
+      {
+        ASSERT_GE(fileContent.size(), static_cast<std::size_t>(l));
+        futures.emplace_back(std::async(std::launch::async, testUploadFromBuffer, c, l));
+        futures.emplace_back(std::async(std::launch::async, testUploadFromFile, c, l));
+      }
+    }
+    for (auto& f : futures)
+    {
+      f.get();
+    }
+  }
+
+  TEST_F(FileShareFileClientTest, ConcurrentDownload)
```

Comment:
i know you copied this from blob test case. But that one is so slow that i was gonna rewrite it.

Reply:
Then I look forward to your rewrites. :)

Reply:
https://github.com/Azure/azure-sdk-for-cpp/pull/513